src/checkers/cwe_476.ml

```diff
@@ -1,47 +1,280 @@
-
-let find_blk_tid_in_sub blk_tid sub =
-  Seq.find (Term.enum blk_t sub) ~f:(fun b -> (Term.tid b) = blk_tid)
-
-let get_jmps blk = Seq.filter (Blk.elts blk) ~f:(fun elt -> match elt with
-          | `Phi phi -> false
-          | `Def def -> false
-          | `Jmp jmp -> true )
-                   |> Seq.map ~f:(fun j -> match j with
-                       | `Jmp jmp -> jmp
-                       | _ -> assert(false))
-
-let jmp_cond_checks_zf jmp =
-  let e = Jmp.cond jmp in
-  (Exp.to_string e) = "~ZF" || (Exp.to_string e) = "ZF"
-
-(* Check if next block contains when zf = 0 goto, if not then there is a chance that this yields a null pointer deref *)
-let check_null_pointer proj prog sub blk jmp tid_map symbols =
-  Seq.iter (Graphs.Tid.Node.succs (Term.tid blk) (Sub.to_graph sub)) ~f:(
-    fun next_blk -> match find_blk_tid_in_sub next_blk sub with
-      | Some b -> begin
-          (* ToDo: Check if there is a definition of ZF = 0 *)
-          let jmps = get_jmps b in
-          match Seq.find jmps ~f:jmp_cond_checks_zf with
-          | Some _ -> ()
-          | None -> Log_utils.warn "[%s] {%s} (NULL Pointer Dereference) There is no check if the return value is NULL at %s (%s)."
-                      name
-                      version
-                      (Address_translation.translate_tid_to_assembler_address_string (Term.tid blk) tid_map)
-                      (Symbol_utils.get_symbol_name_from_jmp jmp symbols)
-          end
-      | _ -> assert(false))
-
-let check_cwe prog proj tid_map symbol_names =
-  match symbol_names with
-  | hd::[] ->
-     let symbols = Symbol_utils.build_symbols hd prog in 
-     let calls = call_finder#run prog [] in
-     let relevant_calls = filter_calls_to_symbols calls symbols in
-     check_calls relevant_calls prog proj tid_map symbols check_null_pointer
-  | _ -> failwith "[CWE476] symbol_names not as expected"
+let version = "0.2"
+
+(* Access type denotes whether a variable var gets accessed or the memory stored at
+   address var gets accessed *)
+type access_type = | Access of Bil.var | MemAccess of Bil.var | NoAccess
+
+(* The union of two accesses is the higher access with MemAcces > Access > NoAccess *)
+let union_access access1 access2 : access_type =
+  match access1 with
+  | MemAccess(_) -> access1
+  | Access(_) -> begin
+      match access2 with
+      | MemAccess(_) -> access2
+      | Access(_) | NoAccess -> access1
+    end
+  | NoAccess -> access2
+
+
+(* union of three accesses for convenience *)
+let union_access_triple access1 access2 access3 =
+  union_access access1 access2
+  |> union_access access3
+
+(* the state contains a list of pairs of register names containing an unchecked
+   return value and the term identifiers of the block where the unchecked
+   return value was generated. *)
+module State = struct
+  type t = (Var.t * Tid.t) list
+
+  (** adds var as a tainted register (with the taint source given by tid) *)
+  let add state var tid =
```

Comment:
I am not sure how to introduce "old_state" and "new_state" without decreasing readability. I could rename State.add to something like State.add_elem, but you probably had something else in mind.

Reply:
I am referring to the name "state" that you use for several variables (e.g. as parameter for the fun, in the let clause)

Reply:
I could rename state to something like tainted_register_list so that its purpose gets more obvious. I just do not think that I should name something "new state", because it would always be the return value and naming return values in ocaml feels unnecessary verbose.